AmplifyPlugins/Geo/Sources/AWSLocationGeoPlugin/Dependency/AWSDeviceTracker.swift

```diff
@@ -35,31 +39,222 @@ class AWSDeviceTracker: NSObject, CLLocationManagerDelegate, DeviceTrackingBehav
-        if options.requestAlwaysAuthorization {
-            locationManager.requestAlwaysAuthorization()
-        }
         locationManager.allowsBackgroundLocationUpdates = options.allowsBackgroundLocationUpdates
         locationManager.pausesLocationUpdatesAutomatically = options.pausesLocationUpdatesAutomatically
         locationManager.activityType = options.activityType
         locationManager.showsBackgroundLocationIndicator = options.showsBackgroundLocationIndicator
         locationManager.distanceFilter = options.distanceFilter
     }
     
-    func startTracking() {
+    func startTracking(for device: Geo.Device) throws {
+        networkMonitor.start()
+        UserDefaults.standard.removeObject(forKey: AWSDeviceTracker.deviceIDKey)
+        UserDefaults.standard.set(device.id, forKey: AWSDeviceTracker.deviceIDKey)
         locationManager.delegate = self
-        locationManager.startUpdatingLocation()
-        if self.wakeAppForSignificantLocationUpdates {
-            locationManager.startMonitoringSignificantLocationChanges()
-        }
+        try checkPermissionsAndStartTracking()
     }
     
     func stopTracking() {
+        // flush out stored events
+        Task {
+            do {
+                try await sendStoredLocationsToService()
+                UserDefaults.standard.removeObject(forKey: AWSDeviceTracker.deviceIDKey)
+            } catch {
+                // TODO: send error on Hub
+            }
+        }
+        
+        networkMonitor.cancel()
         locationManager.stopUpdatingLocation()
         locationManager.stopMonitoringSignificantLocationChanges()
+        locationManager.delegate = nil
+        UserDefaults.standard.removeObject(forKey: AWSDeviceTracker.lastLocationUpdateTimeKey)
+        UserDefaults.standard.removeObject(forKey: AWSDeviceTracker.lastUpdatedLocationKey)
+    }
+    
+    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
+        if let clError = error as? CLError {
+                switch clError {
+                case CLError.denied:
+                    // user denied location services, stop tracking
+                    stopTracking()
+                default:
+                    Amplify.log.error(error: error)
+                }
+            } else {
+                Amplify.log.error(error: error)
+            }
+    }
+    
+    public func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
+        do {
+            try checkPermissionsAndStartTracking()
+        } catch {
+            Amplify.log.error(error: error)
+        }
     }
     
     public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
-        // TODO: Save position locally
-        proxyDelegate?.didUpdateLocations(manager, locations)
+        // if network is unreachable and `disregardLocationUpdatesWhenOffline` is set, don't store locations
```

Comment:
can this function get broken down further so that it is easier to read and maintain?